hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/BucketUtilizationMetrics.java

```diff
@@ -0,0 +1,112 @@
+ * the License.
+ */
+
+package org.apache.hadoop.ozone.om;
+
+import org.apache.hadoop.hdds.annotation.InterfaceAudience;
+import org.apache.hadoop.hdds.utils.db.cache.CacheKey;
+import org.apache.hadoop.hdds.utils.db.cache.CacheValue;
+import org.apache.hadoop.metrics2.MetricsCollector;
+import org.apache.hadoop.metrics2.MetricsInfo;
+import org.apache.hadoop.metrics2.MetricsSource;
+import org.apache.hadoop.metrics2.MetricsSystem;
+import org.apache.hadoop.metrics2.annotation.Metrics;
+import org.apache.hadoop.metrics2.lib.DefaultMetricsSystem;
+import org.apache.hadoop.ozone.OzoneConsts;
+import org.apache.hadoop.ozone.om.helpers.OmBucketInfo;
+
+import java.util.Iterator;
+import java.util.Map.Entry;
+
+/**
+ * A class for collecting and reporting bucket utilization metrics.
+ * <p>
+ * Available metrics:
+ * <ul>
+ *   <li>Bytes used in bucket.
+ *   <li>Bucket quote in bytes.
+ *   <li>Bucket quota in namespace.
+ *   <li>Bucket available space. Calculated from difference between used bytes in bucket and bucket quota.
+ *   If bucket quote is not set then this metric show -1 as value.
+ * </ul>
+ */
+@InterfaceAudience.Private
+@Metrics(about = "Ozone Bucket Utilization Metrics", context = OzoneConsts.OZONE)
+public class BucketUtilizationMetrics implements MetricsSource {
+
+  private static final String SOURCE = BucketUtilizationMetrics.class.getSimpleName();
+
+  private final OMMetadataManager metadataManager;
+
+  public BucketUtilizationMetrics(OMMetadataManager metadataManager) {
+    this.metadataManager = metadataManager;
+  }
+
+  public static BucketUtilizationMetrics create(OMMetadataManager metadataManager) {
+    MetricsSystem ms = DefaultMetricsSystem.instance();
+    return ms.register(SOURCE, "Bucket Utilization Metrics", new BucketUtilizationMetrics(metadataManager));
+  }
+
+  @Override
+  public void getMetrics(MetricsCollector collector, boolean all) {
+    Iterator<Entry<CacheKey<String>, CacheValue<OmBucketInfo>>> bucketIterator = metadataManager.getBucketIterator();
+
+    while (bucketIterator.hasNext()) {
+      Entry<CacheKey<String>, CacheValue<OmBucketInfo>> entry = bucketIterator.next();
+      OmBucketInfo bucketInfo = entry.getValue().getCacheValue();
+      if (bucketInfo == null) {
+        continue;
+      }
+
+      long availableSpace = bucketInfo.getQuotaInBytes() - bucketInfo.getUsedBytes();
+
+      collector.addRecord(SOURCE)
+          .setContext("Bucket metrics")
+          .tag(BucketMetricsInfo.VolumeName, bucketInfo.getVolumeName())
+          .tag(BucketMetricsInfo.BucketName, bucketInfo.getBucketName())
+          .addGauge(BucketMetricsInfo.BucketUsedBytes, bucketInfo.getUsedBytes())
+          .addGauge(BucketMetricsInfo.BucketQuotaBytes, bucketInfo.getQuotaInBytes())
+          .addGauge(BucketMetricsInfo.BucketQuotaNamespace, bucketInfo.getQuotaInNamespace())
+          .addGauge(BucketMetricsInfo.BucketAvailableBytes, availableSpace > 0 ? availableSpace : -1);
```

Comment:
Available space should be minimum 0?
```suggestion
          .addGauge(BucketMetricsInfo.BucketAvailableBytes, Math.max(availableSpace, 0));
```